samcli/lib/build/app_builder.py

```diff
@@ -0,0 +1,209 @@
+
+            if logical_id not in built_artifacts:
+                # this resource was not built. So skip it
+                continue
+
+            # Artifacts are written relative to the output template because it makes the template portability
+            #   Ex: A CI/CD pipeline build stage could zip the output folder and pass to a
+            #   package stage running on a different machine
+            artifact_relative_path = os.path.relpath(built_artifacts[logical_id], target_dir)
+
+            resource_type = resource.get("Type")
+            if resource_type == "AWS::Serverless::Function":
+                template_dict["Resources"][logical_id]["Properties"]["CodeUri"] = artifact_relative_path
+
+            if resource_type == "AWS::Lambda::Function":
+                template_dict["Resources"][logical_id]["Properties"]["Code"] = artifact_relative_path
+
+        return template_dict
+
+    def _build_function(self, function_name, codeuri, runtime):
+
+        capability = self._builder_capabilities[runtime]
+
+        # Create the arguments to pass to the builder
+        code_dir = os.path.normpath(os.path.join(self.source_root, codeuri))
+        # artifacts directory will be created by the builder
+        artifacts_dir = os.path.join(self.build_dir, function_name)
+
+        # TODO: Create this
+        scratch_dir = None
+
+        # TODO: how do we let customers specify where the manifests are? Or change the name of the file?
+        manifest_path = os.path.join(code_dir, capability["manifest_name"])
+
+        if self.container_manager:
+            return self._build_function_on_container(capability,
+                                                     code_dir,
+                                                     artifacts_dir,
+                                                     manifest_path,
+                                                     runtime)
+        else:
+            builder = LambdaBuilder(language=capability["language"],
+                                    dependency_manager=capability["dependency_manager"],
+                                    application_framework=capability["application_framework"])
+
+            # TODO: Add try-catch and raise an internal exception if Workflow failed
+            builder.build(code_dir,
+                          artifacts_dir,
+                          scratch_dir,
+                          manifest_path,
+                          runtime=runtime)
+
+            return artifacts_dir
+
+    def _build_function_on_container(self,
+                                     capability,
+                                     source_dir,
+                                     artifacts_dir,
+                                     manifest_path,
+                                     runtime):
+
+        # If we are printing debug logs in SAM CLI, the builder library should also print debug logs
+        log_level = LOG.getEffectiveLevel()
```

Comment:
Is it possible to set this higher in the flow? This is a common pattern we should be doing (for example SAM).

Reply:
No, this is specifically passed to the builder library running inside a container. So think of this as an equivalent of `--debug` flag passed to this library.